File: leetcode/1382.Balance-a-Binary-Search-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
https://leetcode.com/problems/balance-a-binary-search-tree/description/

Time complexity: O(n), n = len(root)
Space complexity: O(n)
"""
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        vals = []

        def dfs(node: TreeNode):
            if node is None: return
            vals.append(node.val)
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        vals.sort()

        def to_balance_BST(l: int, r: int) -> TreeNode:
            if l > r: return None
            if l == r: return TreeNode(vals[l])
            mid = (r + l) // 2
            cur = TreeNode(vals[mid])
            cur.left = to_balance_BST(l, mid - 1)
            cur.right = to_balance_BST(mid + 1, r)
            return cur

        return to_balance_BST(0, len(vals) - 1)

File: engineering/LeetCode/1382.balance-a-binary-search-tree.ts
/*
 * @lc app=leetcode id=1382 lang=typescript
 *
 * [1382] Balance a Binary Search Tree
 */

import { TreeNode } from './model';

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 * Time complexity: O(n)
 * Space complexity: O(n)
 * @see https://leetcode.com/problems/balance-a-binary-search-tree/discuss/539686/JavaC%2B%2B-Sorted-Array-to-BST-O(N)-Clean-code
 */
function balanceBST(root: TreeNode | null): TreeNode | null {
    /**
     * by given a binary search tree, using in-order can retrieve a sorted array.
     */
    const sortedArray = (() => {
        let self: number[] = [];
        const inOrderTraverse = (node: TreeNode | null) => {
            if (node === null) { return; }
            inOrderTraverse(node.left);
            self.push(node.val);
            inOrderTraverse(node.right);
        }
        inOrderTraverse(root);
        return self;
    })();

    const buildBalancedBST = (l: number, r: number): TreeNode | null => {
        if (l > r) { return null; }
        const mid = Math.floor((r - l) / 2) + l;
        const node = new TreeNode(sortedArray[mid]);
        node.left = buildBalancedBST(l, mid - 1);
        node.right = buildBalancedBST(mid + 1, r);
        return node;
    };

    return buildBalancedBST(0, sortedArray.length - 1);
};
// @lc code=end

